Algorithms_and_Data_Structures/Design_and_Analysis_of_Algorithms/Single_Source_Shortest_path_problems/a_star.py

```diff
@@ -0,0 +1,74 @@
+import heapq
+
+class Graph:
+    def __init__(self):
+        self.adjacency_list = {}
+
+    def add_edge(self, u, v, cost):
+        if u not in self.adjacency_list:
+            self.adjacency_list[u] = []
+        if v not in self.adjacency_list:
+            self.adjacency_list[v] = []
+        self.adjacency_list[u].append((v, cost))
+        self.adjacency_list[v].append((u, cost))  # For undirected graph
+
+def heuristic(a, b):
+    # Example heuristic function (can be replaced with a real heuristic)
+    # Here we use a simple Manhattan distance as an example for 2D coordinates
+    return abs(a[0] - b[0]) + abs(a[1] - b[1])
+
+def a_star_search(graph, start, goal):
+    open_set = []
+    heapq.heappush(open_set, (0, start))  # (f_score, node)
+    
+    came_from = {}
+    g_score = {node: float('inf') for node in graph.adjacency_list}
+    g_score[start] = 0
+    
+    f_score = {node: float('inf') for node in graph.adjacency_list}
+    f_score[start] = heuristic(start, goal)
+
+    while open_set:
+        current = heapq.heappop(open_set)[1]
+
+        if current == goal:
+            # Reconstruct the path
```

Comment:
Construct the path outside the loop since this only happens when the goal is found